promote-release-to-box.py

```diff
@@ -0,0 +1,290 @@
+                rt_path.writeto(out=file_name, progress_func=None)
+                logging.info(f'Download complete.')
+            
+            if not os.path.exists(check_sum_file_name):
+                logging.info(f'Saving check sum file {check_sum_file_name}')
+                with open(check_sum_file_name, 'w') as f:
+                    f.write('sha1='+ sha1 +'\n')
+                    f.write('sha256='+ sha256 +'\n')
+                    f.write('md5='+ md5+'\n')
+                
+        release_artifact_downloaded.append(file_name)
+                
+    logging.info(f'Release artifacts {release_artifact_downloaded}')
+    return release_artifact_downloaded
+
+
+def replace_versions(text, replacement_word):
+    pattern = r'\${.*?}'
+    return re.sub(pattern, replacement_word, text)
+
+
+def get_manifest_yaml(version, manifest_file = 'manifest.yaml'):
+    with open(manifest_file, 'r') as f:
+        file_content = f.read()
+
+    new_file_content = replace_versions(file_content, version)
+    yaml_obj = yaml.safe_load(new_file_content)
+    return yaml_obj
+
+
+def process_manifest_yaml(yaml_data, parent=None):
+    '''
+    Returns a dict with parent as value, child as key.
+    {'pad-ee-9.5.1.0-dist.zip': 'ee/client-tools',
+     'pdi-ee-client-9.5.1.0-dist.zip': 'ee/client-tools',
+     'pentaho-analysis-ee-9.5.1.0-dist.zip': 'ee/client-tools',
+     'pentaho-big-data-ee-package-9.5.1.0-dist.zip': 'ee/client-tools',
+     'pme-ee-9.5.1.0-dist.zip': 'ee/client-tools',
+     'prd-ee-9.5.1.0-dist.zip': 'ee/client-tools',
+     ...}
+    
+    '''
+    result = {}
+    for key, value in yaml_data.items():
+        current_key = key if parent is None else f"{parent}/{key}"
+        if isinstance(value, dict):
+            result.update(process_manifest_yaml(value, parent=current_key))
+        elif isinstance(value, list):
+            for item in value:
+                if '$' not in item or '{' not in item:
+                    result[item] = current_key
+    logging.info(f'Read manifest {result}')
+    return result
+
+
+def get_manifest_buildinfo_intersect(file_folder_dict, builds_output_json):
+    d = {}
+    manifest_files = set(file_folder_dict.keys()).intersection(set([artifact['path'].split('/')[-1] for artifact in builds_output_json]))
+    manifest_files
+    for file_to_release in manifest_files:
+        d[file_to_release] = file_folder_dict[file_to_release]
+        d[file_to_release + '.sum'] =  file_folder_dict[file_to_release]
+    return d
+
+
+def set_box_client(client_id, client_secret, box_subject_id):
+    token=generate_access_token(client_id,client_secret,box_subject_id)
+    oauth = OAuth2(
+        client_id=client_id,
+        client_secret=client_secret,
+        access_token=token,  
+    )
+
+    return Client(oauth)
+
+
+def upload_one_artifact_to_box(folder_id, file_name, client):
+
+    with open(file_name, 'rb') as file:
+        box_file = client.folder(folder_id).upload_stream(file, file_name)
```

Comment:
Don't remember exactly, but there was some issue if the file already existed in box... not sure if we would have that issue here in this use case.

Reply:
Haven't tried this yet, will update you after tried.

Reply:
@smaring might have more input on uploading a file to box when one exist there already.

Reply:
```
boxsdk.exception.BoxAPIException: Message: Item with the same name already exists
Status: 409
Code: item_name_in_use
Request ID: 4t0o8ghg89rzuqgy
Headers: {'Server': 'nginx', 'Date': 'Mon, 24 Jul 2023 12:32:48 GMT', 'Content-Type': 'application/json', 'Content-Length': '481', 'cache-control': 'no-cache, no-store', 'x-envoy-upstream-service-time': '2331', 'strict-transport-security': 'max-age=31536000', 'X-Box-Original-Ingress-ADC-Host': 'prod-c-traffic-manager-dz94', 'Via': '1.1 google', 'Alt-Svc': 'h3=":443"; ma=2592000,h3-29=":443"; ma=2592000'}
URL: https://upload.box.com/api/2.0/files/content
Method: POST
Context Info: {'conflicts': {'type': 'file', 'id': '1263740384766', 'file_version': {'type': 'file_version', 'id': '1380881600766', 'sha1': '7bfae4310b2331c1f616697d5336cad08dc505bb'}, 'sequence_id': '0', 'etag': '0', 'sha1': '7bfae4310b2331c1f616697d5336cad08dc505bb', 'name': 'xxx.zip'}}
```

I remember now, it will pop 409 error and the script will stop. There was a try catch block there which I deleted it since having a exception block in the subroutine will need to trigger interrupt exception X number of times for X number of files that there are to cancel the entire pipeline, thus I removed it instead.